### DISCUSSION TITLE
Spawning entities with passengers on startup

### DISCUSSION THREAD

**Discussion by MrDienns** [Minestom/Minestom#1749] [Help]
Hey all,

I'm currently trying to spawn a series of entities as soon as the server starts up. These entities contain multiple invisible enderpearls as passengers, which we use for nametags. The below code works (see first screenshot) when spawning them *after* I've joined the server and ran a command, but it does not work properly doing this on startup (so, before I've logged in). I'm unsure why.

```kotlin
    override suspend fun onCreate(event: NPCCreatedEvent) {
        val instance = MinecraftServer.process().instance().instances.first()
        entity.setInstance(instance, event.location.toPosition()).thenAccept {
            initializeTags(instance)
        }
    }

    private fun initializeTags(instance: Instance) {
        
        titleTag = Entity(EntityType.ENDER_PEARL)
        titleTag.isCustomNameVisible = true
        titleTag.customName = Component.text("NPC", NamedTextColor.GREEN)

        entity.addPassenger(titleTag)

        nameTag = Entity(EntityType.ENDER_PEARL)
        nameTag.isCustomNameVisible = true
        nameTag.customName = Component.text(model.name)

        titleTag.addPassenger(nameTag)
    }
```

However, doing the same thing on startup via the scheduler, will make it so that the enderpearls hit the floor as soon as I log in (see 2nd screenshot). Running the command manually still works fine, *after* I've joined the server.

I've tried to change the code up a bit, by setting the instance of the enderpearls and using the completable future to spawn them one by one, but that doesn't work either. See below.

```kotlin
    override suspend fun onCreate(event: NPCCreatedEvent) {
        val instance = MinecraftServer.process().instance().instances.first()
        entity.setInstance(instance, event.location.toPosition()).thenAccept {
            initializeTags(instance)
        }
    }

    private fun initializeTags(instance: Instance) {

        titleTag = Entity(EntityType.ENDER_PEARL)
        titleTag.isCustomNameVisible = true
        titleTag.customName = Component.text("NPC", NamedTextColor.GREEN)
        titleTag.setInstance(instance).thenAccept {

            entity.addPassenger(titleTag)
            
            nameTag = Entity(EntityType.ENDER_PEARL)
            nameTag.isCustomNameVisible = true
            nameTag.customName = Component.text(model.name)
            nameTag.setInstance(instance).thenAccept {

                titleTag.addPassenger(nameTag)
            }
        }
    }
```

It gives the same result as the other option (see screenshot 3). However, running it manually via a command now also doesn't work anymore as the enderpearls are not added to the pig entity anymore. Anyone know what's going on or what the best practices are here?

![image](https://user-images.githubusercontent.com/15938901/219668620-bbd9d81f-9482-47a1-8937-7548c634df29.png)
![image](https://user-images.githubusercontent.com/15938901/219668647-9f46d862-30ce-428e-8845-e64e456bf686.png)
![image](https://user-images.githubusercontent.com/15938901/219668678-9942ebdb-4da9-4c0c-ad84-f74830e030a8.png)

**Comment by mworzala** [comment]
Minestom doesn't simulate ender pearls moving, so I guess this is the client doing it. I cannot test it at the moment, but perhaps passengers aren't correctly sent on join. If you set the ender pearls to have no gravity, do they stay in place?

**Reply by MrDienns** [reply]
Disabling gravity on the enderpearls shows some interesting results. I've also given an AI goal to the pig to run around a bit, as it may yield more information.

Upon joining the server, the enderpearls were in a different location from where the NPC spawns (likely where the pig was at the moment of me joining);

![image](https://user-images.githubusercontent.com/15938901/219672883-dbd8ef55-0016-4766-846d-67a5370b929f.png)

Running around a bit, at some point they will teleport towards the location of the mob. Seems to be every minute.

![image](https://user-images.githubusercontent.com/15938901/219673032-fe2a4379-2d9e-4c83-a5a6-a1105a1bc871.png)

**Reply by mworzala** [reply]
I have no idea why they teleport, but it seems that indeed minestom isn't sending passenger information correctly when joining.

**Reply by MrDienns** [reply]
Thanks for taking a look. I've created a bug here; https://github.com/Minestom/Minestom/issues/1750
